src/renderers/shared/fiber/ReactFiberUpdateQueue.js

```diff
@@ -399,24 +214,50 @@ function beginUpdateQueue(
     queue.isProcessing = true;
   }
 
-  // Calculate these using the the existing values as a base.
-  let callbackList = queue.callbackList;
-  let hasForceUpdate = queue.hasForceUpdate;
+  // Reset the remaining expiration time. If we skip over any updates, we'll
+  // increase this accordingly.
+  queue.expirationTime = NoWork;
 
-  // Applies updates with matching priority to the previous state to create
-  // a new state object.
-  let state = prevState;
+  let state = queue.baseState;
   let dontMutatePrevState = true;
   let update = queue.first;
-  while (update !== null && update.expirationTime <= renderExpirationTime) {
-    // Remove each update from the queue right before it is processed. That way
-    // if setState is called from inside an updater function, the new update
-    // will be inserted in the correct position.
-    queue.first = update.next;
-    if (queue.first === null) {
-      queue.last = null;
+  let didSkip = false;
+  while (update !== null) {
+    const updateExpirationTime = update.expirationTime;
+    if (
+      updateExpirationTime === NoWork ||
```

Comment:
What does it mean for an update to have an expiration time of NoWork? That should never happen, right?

Reply:
No but I feel a pang of guilt whenever I intentionally don't handle a case.  If this were Reason I'd be forced to handle it :D